### ISSUE TITLE
let's separate the MD table string generation functionality into a standalone library

### ISSUE THREAD

**Issue by revelt** [jfmengels/all-contributors-cli#67]
hi all,

Currently, if somebody wants to make an alternative tool that generates the same-looking contributors table, there's no way to automatically ensure that generated MD string (table) matches the "official" format of the contributors table.

Ideally, whatever CLI does - reading its configs, parsing existing table and so on should lead up to generating some sort of a data structure (object). It would then be fed into a separate library which renders the MD table string from that data structure.

This way, anybody who doesn't like the CLI behaviour or want completely different tool to maintain the contributors could still generate compatible contributors' table. In turn, that generated readme would still be compatible with this CLI or any other tools from `all-contributors` family.

At the moment, I can easily reference the format of the contributors MD table and write my own renderer, but as #54 showed, the format of the table evolves in time and separation of CLI from renderer would help to keep everybody in sync and foster the **whole ecosystem of tools** that can generate the **standard** `all-contributors` table.

So that was `renderer`.

A `parser` would be nice too - an "official" parser which can parse the readme and extract info from `all-contributors` table **according to official, latest spec**. Let's say somebody wants to update the contributors badge according to existing contents of readme and its contributors table.

**Comment by machour** [comment]
Hi @revelt,

For me, the `all-contributors` specification isn't about the format of the table, it's about the known contributions type, and reminding maintainers to give credit where it's due. The specification is in another repo: [all-contributors](https://github.com/kentcdodds/all-contributors)

`all-contributors-cli` is just a tool that help keeping up with contributions, and that additionally generates a nice table for you. You can even customize the rendering, so there's pretty much nothing official (eg, specification) about this tool.

And as rendering can be customized, I don't see a (easy) way to parse the generated markdown.

In other words, anyone can write his own tool and maintain/render contributors the way he wants, this tool simply implements one of the ways to do that.

/cc @kentcdodds to correct me if my vision of things is wrong.

**Comment by revelt** [comment]
@machour I understand your point. Yes, `all-contributors` spec is a different thing. I'm looking for a different spec. Spec for "default" table generation.

Do you agree that `all-contributors-cli` has a **a default preset** among its code, against which it generates the default table (if there are no customisations)? I'd like to tap this default preset (string generator function) in a shape of external lib. Because _there is_ one, and it should be separate from the rest of CLI code...

As I mentioned #54 added bold around labels, for example, and in such way, mutated the "default" table preset. If other tools would like to generate contrib tables, compatible with this tool, that would break compatibility. Tables from other tools would generate contrib tables without `b` tags (using #54 case as example) until it was updated.

Separating CLI **into parts** would promote the **ecosystem of compatible apps** that generate the same "default" contrib table style.

Do you see my point?

**Comment by machour** [comment]
@revelt I understand your point, but again, for me, there's nothing official about `all-contributors-cli`, nor about the default table it generates. 
I personally don't like the idea of splitting it across different parts. It's a simple tool and should stay as simple as possible (this is my opinionated opinion).

Do you know of any other tool that generates a contributor list based on the spec? Would you like to see some changes in this tool?

**Comment by revelt** [comment]
> Do you know of any other tool that generates a contributor list based on the spec?
I'm about to release one. Before I do release it I'd be keen to tap the "official" string generator so it's output is guaranteed to be compatible with this CLI, hence me pestering you guys here.

> Would you like to see some changes in this tool?
No. Having said that, I prefer to keep contributors in `package.json` and have no all-contributors config at all. I'm using `format-package` so I have no trouble to visually-grep through the package.json contents.

**Comment by kentcdodds** [comment]
Feel free to do whatever you like with the format and stuff. I don't care to much about the format. And I don't think it's worth trying to standardize.

Thanks for building a tool! Make sure to make a pull request to the spec repo when you're done! 👏

**Comment by revelt** [comment]
An update. 

After github selling us out, I moved all my libraries to Bitbucket. But it doesn't support HTML in readme's! This also means, we can't use Prettier ignore blocks and Prettier mangles both doctoc TOC and Contributor's tables.

I'll research ways to generate Bitbucket Contributors tables, but first need to complete the migration from Github.

Either way, now I believe in readme maintenance automation and processing it comprehensively. Just a single CLI is not a solution, this task is only a piece of a larger puzzle.

So, I'll close; thank you everybody for attention 👍
